GitUI/BranchTreePanel/MenuItemsGenerator.cs

```diff
@@ -0,0 +1,124 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Windows.Forms;
+using GitUI.BranchTreePanel.Interfaces;
+
+namespace GitUI.BranchTreePanel
+{
+    internal class MenuItemsGenerator<TNode> : IMenuItemsGenerator<TNode> where TNode : class, INode
+    {
+        private readonly IMenuItemFactory _menuItemFactory;
+
+        private readonly Dictionary<MenuItemKey, ToolStripItem> _itemsIndex;
+
+        private readonly List<ToolStripItemWithKey> _items;
+
+        public MenuItemsGenerator(IMenuItemFactory menuItemFactory)
+        {
+            _menuItemFactory = menuItemFactory;
+            _items = new List<ToolStripItemWithKey>();
+            _itemsIndex = new Dictionary<MenuItemKey, ToolStripItem>();
+            Generate();
+        }
+
+        private void Generate()
+        {
+            AddRange(CreateGitRefItems());
+            AddRange(CreateRename());
+            AddRange(CreateDelete());
+        }
+
+        #region IEnumerable
+        public IEnumerator<ToolStripItemWithKey> GetEnumerator()
+        {
+            return _items.GetEnumerator();
+        }
+
+        IEnumerator IEnumerable.GetEnumerator()
+        {
+            return GetEnumerator();
+        }
+        #endregion
+
+        public bool TryGetMenuItem(MenuItemKey key, out ToolStripItem item)
+        {
+            return _itemsIndex.TryGetValue(key, out item);
+        }
+
+        private void AddRange(IEnumerable<ToolStripItemWithKey> createResults)
+        {
+            var array = createResults.ToArray();
+            _items.AddRange(array);
+            array.ForEach(r => _itemsIndex.Add(r.Key, r.Item));
```

Comment:
```suggestion
            _items.ForEach(r => _itemsIndex[r.Key] = r.Item));
```

Reply:
_items is already containg items, so it would re-add old items and crash here 🙈